torch/csrc/jit/autodiff.cpp

```diff
@@ -285,6 +293,8 @@ static ReverseDetails addReverseInline(Gradient& grad_desc,
     auto inputs = node->inputs();
     if (std::none_of(inputs.begin(), inputs.end(), requires_grad))
       continue;
+    if (outputNotRequiresGrad(node)) continue;
+    if (node->kind() == aten::type_as && !requires_grad(inputs[0])) continue;
```

Comment:
This logic is repeated, but it's pretty tricky, and it's easy to forget to add a case in one of those two places if we'll hit other edges cases. Can you isolate it into a function? Something like
```cpp
bool outputRequiresGrad(Node* node, std::function<bool(Value*)> requires_grad) {
  switch (node->kind()) {
    case aten::le:
    case aten::ge:
    ...
    case aten::eq:
      return false;
    case aten::type_as:
      return requires_grad(inputs[0]);
    default:
      return std::any_of(node->inputs().begin(), node->inputs().end(), requires_grad);
  }
}
```